pkg/controller/dataimportcron-controller.go

```diff
@@ -0,0 +1,594 @@
+			UID:        dataImportCron.UID,
+		})
+	}
+	sourcePVC := &cdiv1.DataVolumeSourcePVC{
+		Namespace: dataImportCron.Namespace,
+		Name:      dataImportCron.Status.CurrentImportDataVolumeName,
+	}
+	if dataSource.Spec.Source.PVC == nil || *dataSource.Spec.Source.PVC != *sourcePVC {
+		dataSource.Spec.Source.PVC = sourcePVC
+		if err := r.client.Update(ctx, dataSource); err != nil {
+			log.Error(err, "Unable to update DataSource with source PVC", "Name", dataSourceName, "PVC", sourcePVC)
+			return err
+		}
+	}
+	return nil
+}
+
+func (r *DataImportCronReconciler) updateDataImportCronOnSuccess(ctx context.Context, dataImportCron *cdiv1.DataImportCron) error {
+	sourcePVC := &cdiv1.DataVolumeSourcePVC{
+		Namespace: dataImportCron.Namespace,
+		Name:      dataImportCron.Status.CurrentImportDataVolumeName,
+	}
+	if dataImportCron.Status.LastImportedPVC == nil || *dataImportCron.Status.LastImportedPVC != *sourcePVC {
+		dataImportCron.Status.LastImportedPVC = sourcePVC
+		now := metav1.Now()
+		dataImportCron.Status.LastImportTimestamp = &now
+	}
+	// Mark no current import
+	dataImportCron.Status.CurrentImportDataVolumeName = ""
+	return nil
+}
+
+func (r *DataImportCronReconciler) createImportDataVolume(ctx context.Context, dataImportCron *cdiv1.DataImportCron) error {
+	log := r.log.WithName("createImportDataVolume")
+	dataSourceName := dataImportCron.Spec.ManagedDataSource
+	digest := dataImportCron.Annotations[AnnSourceDesiredDigest]
+	dvName := createDvName(dataSourceName, digest)
+	log.Info("Creating new source DataVolume", "name", dvName)
+	dv := newSourceDataVolume(dataImportCron, dvName)
+	if err := r.client.Create(ctx, dv); err != nil {
+		log.Error(err, "Unable to create DataVolume", "name", dvName)
+		return err
+	}
+	// Update references to current import
+	dataImportCron.Status.CurrentImportDataVolumeName = dvName
+	dataImportCron.Status.CurrentImportDigest = digest
+	dataImportCron.Annotations[AnnSourceUpdatePending] = "false"
+	return nil
+}
+
+func (r *DataImportCronReconciler) garbageCollectOldImports(ctx context.Context, dataImportCron *cdiv1.DataImportCron) error {
+	log := r.log.WithName("garbageCollectOldImports")
+	if dataImportCron.Spec.GarbageCollect == nil || *dataImportCron.Spec.GarbageCollect != cdiv1.DataImportCronGarbageCollectOutdated {
+		return nil
+	}
+	selector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			labelDataImportCronName: dataImportCron.Name,
+		},
+	})
+	if err != nil {
+		return err
+	}
+	dvList := &cdiv1.DataVolumeList{}
+	if err := r.client.List(ctx, dvList, &client.ListOptions{Namespace: dataImportCron.Namespace, LabelSelector: selector}); err != nil {
+		log.Error(err, "error listing dvs")
+		return err
+	}
+	if len(dvList.Items) <= recentImportsToKeepPerCronJob {
+		return nil
+	}
+	sort.Slice(dvList.Items, func(i, j int) bool {
+		return dvList.Items[i].CreationTimestamp.Time.After(dvList.Items[j].CreationTimestamp.Time)
```

Comment:
Shouldn't this function return true if element at i < element at j?  Isn't this the opposite?

Reply:
ah I see now `for _, dv := range dvList.Items[recentImportsToKeepPerCronJob:]`

Reply:
yep, you got it :)